### DISCUSSION TITLE
Best Practice: Flat File vs Database

### DISCUSSION THREAD

**Discussion by Mushoz** [PrivateBin/PrivateBin#1053] [Q&A] [answered]
Normally for most production systems, a database is preferred over flat file storage. However, since PrivateBin basically stores encrypted blobs, I am not so sure if that holds true for PrivateBin as well.

What would you say is best practice for a production privatebin instance? Should I go for a Flat file or Database backend? Curious to your opinions!

**Comment by elrido** [comment, marked as answer]
Given that we now include a migration script, to switch between backends, should you outgrow them, you are now less locked into one of the solutions. My advice would be along these lines:

- If you want to host the instance on a budget, on a small VM or an older model of single board computer at home, consider the Filesystem storage, as it has the least dependencies and will run with little resources. The privatebin.net demo instance uses it and runs on a 512 MiB of RAM VM. It certainly lends itself more to single instance setups, though could be scaled out, with the use of NFS or similar networked filesystems.
- If you want a redundant setup or be able to scale out horizontally in the future, the database backend is probably simpler than setting up a shared, networked, filesystem solution. Except...
- If you are using a cloud provider, you probably prefer to use S3 or GCS. Those backends are also likely cheaper then their managed database offerings, while still letting you scale out the instances.

There were also some prior discussions on this subject:
- [Are you using the database storage and what benefits do you gain from it? #461](https://github.com/PrivateBin/PrivateBin/issues/461)
- [assumptions on the performance of the different storage backends](https://github.com/PrivateBin/PrivateBin/discussions/1027#discussioncomment-4382514)
